projects/vernon_continual/experiments/si.py

```diff
@@ -0,0 +1,179 @@
+# Numenta Platform for Intelligent Computing (NuPIC)
+# Copyright (C) 2021, Numenta, Inc.  Unless you have an agreement
+# with Numenta, Inc., for a separate license for this software code, the
+# following terms and conditions apply:
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+# See the GNU Affero Public License for more details.
+#
+# You should have received a copy of the GNU Affero Public License
+# along with this program.  If not, see http://www.gnu.org/licenses.
+#
+# http://numenta.org/licenses/
+
+"""
+Experiment file that runs Synaptic Intelligence (Zenke, Poole, Ganguli (2017)) in a
+continual learning experiment on permutedMNIST.
+"""
+
+import os
+from copy import deepcopy
+
+import numpy as np
+import ray.tune as tune
+import torch
+import torch.nn.functional as F
+from torch import nn
+
+from nupic.research.frameworks.pytorch.datasets import PermutedMNIST
+from nupic.research.frameworks.vernon import ContinualLearningExperiment, mixins
+
+
+class SIContinualLearningExperiment(mixins.SynapticIntelligence,
+                                    mixins.PermutedMNISTTaskIndices,
+                                    ContinualLearningExperiment):
+    pass
+
+
+def train_si_model(model, loader, optimizer, device, criterion=F.cross_entropy,
+                   post_batch_callback=None, **kwargs):
+    model.train()
+    for batch_idx, (data, target) in enumerate(loader):
+        data = data.flatten(start_dim=1)
+
+        # Share labels between tasks
+        # TODO don't hardcode num classes per task
```

Comment:
where is num classes hardcoded in that function ?

Reply:
On the next line: `target = target % 10`. This assumes there are 10 classes per task, but not a big deal for now. I placed a 'TODO' here so that we can modify this during later code enhancements.